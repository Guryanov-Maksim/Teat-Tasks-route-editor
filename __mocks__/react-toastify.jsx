import React from 'react';

const toast = jest.fn();
toast.dismiss = jest.fn();
toast.update = jest.fn();
toast.TYPE = {
  ERROR: 'error',
};

const ToastContainer = () => <div className="ToastContainer" />;

export { toast, ToastContainer };
