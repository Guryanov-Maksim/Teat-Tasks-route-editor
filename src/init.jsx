import React from 'react';
import { Provider } from 'react-redux';
import i18n from 'i18next';
import { initReactI18next, I18nextProvider } from 'react-i18next';
import * as yup from 'yup';

// import store from './store.js';
import configureAppStore from './store.js';
import { ru, errors } from './locales/index.js';
import App from './components/App.jsx';

const defaultLanguage = 'ru';

export default async () => {
  yup.setLocale(errors);

  const i18nInstance = i18n.createInstance();

  await i18nInstance
    .use(initReactI18next)
    .init({
      lng: defaultLanguage,
      debug: process.env.NODE_ENV === 'development',
      resources: {
        ru,
      },
    });

  const store = configureAppStore();

  return (
    <Provider store={store}>
      <I18nextProvider i18n={i18nInstance}>
        <App />
      </I18nextProvider>
    </Provider>
  );
};
