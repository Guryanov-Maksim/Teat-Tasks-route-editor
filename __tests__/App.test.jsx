import '@testing-library/jest-dom';
import { render, screen, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { toast } from 'react-toastify';

import init from '../src/init.jsx';
// *************************************************************
// TODO

// 1. Realize a check of expect(input).toHaveAttribute('readonly') while form is submitting
//
// 2. Find out why withYMaps.mockImplementationOnce(() => { ... }) dosn't work in tests:
//
// import { withYmaps } from 'react-yandex-maps';
//
// withYMaps
//   .mockImplementationOnce((Component) => {
//       console.log('new withMock implementation called');
//       return () => <Component ymaps={new MockRejectedYmaps()} />;
//     });
// test('network error message', async () => {
//   console.log(withYMaps);
//   withYMaps
//     .mockImplementationOnce((Component) => {
//       console.log('new withMock implementation called');
//       return () => <Component ymaps={new MockRejectedYmaps()} />;
//     });

//   const vdom = await init();
//   render(vdom);
// });
//
// 3. Realize a dragend event for placemarks instead of await user.click(placemark);
//
// 4. Test points reordering when a point was dragged in the points list
//
// *************************************************************

jest.mock('react-beautiful-dnd');
jest.mock('react-yandex-maps');
jest.mock('react-toastify');

const mockToastId = 'toastId';
toast.mockReturnValue(mockToastId);

afterEach(cleanup);

test('should render correctly', async () => {
  const vdom = await init();
  const { container } = render(vdom);
  expect(container.firstChild).toMatchSnapshot();
  const input = screen.getByTestId('new-message');
  expect(input).toBeInTheDocument();
  expect(input).toHaveFocus();
  expect(input).toHaveValue('');
});

test('add point process', async () => {
  const vdom = await init();
  render(vdom);
  const user = userEvent.setup();
  const input = screen.getByTestId('new-message');

  await user.keyboard('{Enter}');

  const feedBack = await screen.findByText('Обязательное поле');
  expect(feedBack).toBeInTheDocument();

  await user.type(input, 'first new point');
  await user.keyboard('{Enter}');

  const newAddress = await screen.findByText('first new point address');
  expect(newAddress).toBeInTheDocument();

  await user.type(input, 'second new point');
  await user.keyboard('{Enter}');

  const newAddress2 = await screen.findByText('second new point address');
  expect(newAddress2).toBeInTheDocument();

  await user.type(input, 'nonexistent address');
  await user.keyboard('{Enter}');

  const feedBack3 = await screen.findByText('Указанный вами адрес не найдет');
  expect(feedBack3).toBeInTheDocument();

  await user.type(input, 'first new point');
  await user.keyboard('{Enter}');

  const networkError = await screen.findByText('Ошибка сети');
  expect(networkError).toBeInTheDocument();
});

test('remove point', async () => {
  const vdom = await init();
  render(vdom);
  const user = userEvent.setup();
  const input = screen.getByTestId('new-message');

  await user.type(input, 'first new point');
  await user.keyboard('{Enter}');

  await user.type(input, 'second new point');
  await user.keyboard('{Enter}');

  const newAddress1 = await screen.findByText('first new point address');
  const newAddress2 = await screen.findByText('second new point address');

  const [firstRemovingButton] = await screen.findAllByTestId('removing-button');

  await user.click(firstRemovingButton);

  expect(newAddress1).not.toBeInTheDocument();
  expect(newAddress2).toBeInTheDocument();
});

test('should update address when a placemark was moved', async () => {
  const vdom = await init();
  render(vdom);

  const user = userEvent.setup();

  const input = screen.getByTestId('new-message');

  await user.type(input, 'Москва');
  await user.keyboard('{Enter}');

  const placemark = await screen.findByText('placemark');

  await user.click(placemark); // simulate a dragend event of a placemark, see TODO above

  expect(toast).toHaveBeenCalled();

  expect(toast.mock.calls[0][0]).toBe('Подождите, идет обновление адреса...');
  expect(toast.dismiss).toHaveBeenCalled();
  expect(toast.dismiss.mock.calls[0][0]).toBe(mockToastId);

  await user.click(placemark); // simulate a dragend event of a placemark, see TODO above
  expect(toast.update).toHaveBeenCalled();
  expect(toast.update.mock.calls[0][0]).toBe(mockToastId);
});
