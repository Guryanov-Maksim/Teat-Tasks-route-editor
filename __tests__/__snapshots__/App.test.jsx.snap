// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  class="h-100 py-3 d-flex flex-column container-fluid"
>
  <h1
    class="text-center mb-4"
  >
    Редактор маршрута
  </h1>
  <div
    class="h-100 overflow-hidden"
  >
    <div
      class="h-100 flex-nowrap pt-1 row"
    >
      <div
        class="h-100 d-flex flex-column col-md-4 col-sm-4 col-6"
      >
        <div
          class="mt-auto px-1 mb-5"
        >
          <form
            class=""
          >
            <div
              class="position-relative col-sm"
            >
              <div
                class="form-floating"
              >
                <input
                  class="form-control"
                  data-testid="new-message"
                  name="address"
                  type="text"
                />
                <div
                  class="invalid-tooltip"
                >
                  errors.null
                </div>
                <label>
                  Введите новую точку
                </label>
              </div>
            </div>
          </form>
        </div>
        <h3
          class="text-center"
        >
          Точки маршрута
        </h3>
        <div
          class="h-100 overflow-auto"
        >
          <div
            class="list-group list-group-flush"
          />
        </div>
      </div>
      <div
        class="h-100 col-md-8 col-sm-8 col-12"
      >
        <div
          class="d-flex h-100 card"
        >
          <div
            class="map"
          >
            <div
              class="route"
            />
          </div>
          <div
            class="ToastContainer"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
